File: frontend/src/components/prs/prMeasures.test.js
import { describe, expect, it } from 'vitest';
import {
  PR_MEASURE_OPTIONS,
  formatPrMeasure,
  measureEntry,
  measureUnavailableCaption,
  prMeasureSpec,
} from './prMeasures';

const loaded = {
  exerciseId: 1,
  exerciseName: 'Bench Press',
  best: { weight: 185, reps: 5, unit: 'lb', est1rm: 208, sessionStartedAt: '2026-07-01T00:00:00Z' },
  measures: {
    heaviest: { value: 225, weightLb: 225, reps: 1, sessionStartedAt: '2026-07-09T00:00:00Z' },
    sessionVolume: { value: 4625, weightLb: null, reps: null, sessionStartedAt: '2026-07-01T00:00:00Z' },
    bestSetVolume: { value: 925, weightLb: 185, reps: 5, sessionStartedAt: '2026-07-01T00:00:00Z' },
    totalReps: { value: 25, weightLb: null, reps: null, sessionStartedAt: '2026-07-01T00:00:00Z' },
  },
  bodyweightOnly: false,
  durationTracked: false,
};

const pullUp = {
  exerciseId: 2,
  exerciseName: 'Pull-Up',
  best: { weight: 0, reps: 12, unit: 'lb', est1rm: 0, sessionStartedAt: '2026-07-02T00:00:00Z' },
  measures: {
    heaviest: null,
    sessionVolume: null,
    bestSetVolume: null,
    totalReps: { value: 40, weightLb: null, reps: null, sessionStartedAt: '2026-07-02T00:00:00Z' },
  },
  bodyweightOnly: true,
  durationTracked: false,
};

const plank = {
  exerciseId: 3,
  exerciseName: 'Plank',
  best: { weight: 0, reps: 0, durationSeconds: 130, unit: 'lb', est1rm: null, sessionStartedAt: '2026-07-04T00:00:00Z' },
  measures: { heaviest: null, sessionVolume: null, bestSetVolume: null, totalReps: null },
  bodyweightOnly: true,
  durationTracked: true,
};

describe('PR measure options', () => {
  // The board's picker and the Trends chart's switcher are the same five words, read off the same
  // specs. If this ever diverges, "Volume" means different things on two screens.
  it('offers exactly the five the exercise chart plots, in the same order', () => {
    expect(PR_MEASURE_OPTIONS.map((o) => o.value)).toEqual([
      'est1rm',
      'heaviest',
      'sessionVolume',
      'bestSetVolume',
      'totalReps',
    ]);
    expect(PR_MEASURE_OPTIONS.map((o) => o.label)).toEqual([
      'Est. 1RM',
      'Top weight',
      'Volume',
      'Best set',
      'Reps',
    ]);
  });

  it('falls back rather than throwing on an unknown measure', () => {
    expect(prMeasureSpec('nonsense').label).toBe('Est. 1RM');
    expect(prMeasureSpec(undefined).label).toBe('Est. 1RM');
  });
});

describe('measureEntry', () => {
  it('reads est. 1RM off `best`, which is where the server keeps it', () => {
    expect(measureEntry(loaded, 'est1rm')).toMatchObject({ value: 208, weightLb: 185, reps: 5 });
  });

  // comparableLb's weight-0 substitution, mirrored client-side: Epley collapses to 0 at weight 0,
  // so every bodyweight PR would tie forever.
  it('ranks a bodyweight lift by reps under est. 1RM', () => {
    expect(measureEntry(pullUp, 'est1rm')).toMatchObject({ value: 12, weightLb: 0, reps: 12 });
  });

  it('ranks a hold by seconds under est. 1RM', () => {
    expect(measureEntry(plank, 'est1rm')).toMatchObject({ value: 130, durationSeconds: 130 });
  });

  // est1rm arrives in the SET's own unit, unlike row.measures which the server normalizes. Without
  // this a kg household would compare 100 against 200 numerically.
  it('normalizes a kg est. 1RM to pounds so mixed-unit history ranks correctly', () => {
    const metric = {
      best: { weight: 100, reps: 5, unit: 'kg', est1rm: 116, sessionStartedAt: '2026-07-01T00:00:00Z' },
    };
    expect(measureEntry(metric, 'est1rm').value).toBeCloseTo(116 * 2.20462, 3);
  });

  it('reads the other four straight off measures', () => {
    expect(measureEntry(loaded, 'heaviest')).toMatchObject({ value: 225, reps: 1 });
    expect(measureEntry(loaded, 'sessionVolume')).toMatchObject({ value: 4625, weightLb: null });
    expect(measureEntry(loaded, 'bestSetVolume')).toMatchObject({ value: 925, weightLb: 185 });
    expect(measureEntry(loaded, 'totalReps')).toMatchObject({ value: 25 });
  });

  it('returns null -- not zero -- for a measure the exercise cannot carry', () => {
    expect(measureEntry(pullUp, 'heaviest')).toBeNull();
    expect(measureEntry(pullUp, 'bestSetVolume')).toBeNull();
    expect(measureEntry(plank, 'totalReps')).toBeNull();
  });

  it('keeps reps available for a bodyweight lift, its one honest record', () => {
    expect(measureEntry(pullUp, 'totalReps')).toMatchObject({ value: 40 });
  });

  // resilience.md axis D: a cache entry written before this shipped has no `measures` key at all.
  it('degrades a row cached before measures existed instead of throwing', () => {
    const legacy = { best: loaded.best };
    expect(measureEntry(legacy, 'heaviest')).toBeNull();
    expect(measureEntry(legacy, 'est1rm')).toMatchObject({ value: 208 });
  });

  it('falls back to the default measure for an unknown key', () => {
    expect(measureEntry(loaded, 'nonsense')).toMatchObject({ value: 208 });
  });
});

describe('formatPrMeasure', () => {
  it('names the set behind a best-set record', () => {
    const entry = measureEntry(loaded, 'bestSetVolume');
    expect(formatPrMeasure(entry, 'bestSetVolume', loaded, 'lb')).toEqual({
      value: '925 lb',
      caption: '185lb×5',
    });
  });

  // The distinction the ? copy exists to make: nothing else on the row says whether a number is one
  // set or a whole session.
  //
  // "One session" is now the FALLBACK for a row with no breakdown -- a cache entry written before
  // that shipped (resilience.md axis D). The fixtures in this block carry no `sets`, so they
  // exercise exactly that path.
  it('says "One session" for the two session totals, and only those', () => {
    expect(formatPrMeasure(measureEntry(loaded, 'sessionVolume'), 'sessionVolume', loaded, 'lb').caption).toBe(
      'One session',
    );
    expect(formatPrMeasure(measureEntry(loaded, 'totalReps'), 'totalReps', loaded, 'lb').caption).toBe(
      'One session',
    );
    expect(formatPrMeasure(measureEntry(loaded, 'heaviest'), 'heaviest', loaded, 'lb').caption).not.toBe(
      'One session',
    );
  });

  // The value IS the weight for Top weight, so repeating it would say nothing -- the reps are the
  // new information.
  it('qualifies top weight with the reps it was lifted for', () => {
    expect(formatPrMeasure(measureEntry(loaded, 'heaviest'), 'heaviest', loaded, 'lb')).toEqual({
      value: '225 lb',
      caption: '× 1',
    });
  });

  it('names the record instead of reps for a loaded hold, whose reps are always 0', () => {
    const loadedPlank = { ...plank, measures: { ...plank.measures, heaviest: { value: 25, weightLb: 25, reps: 0, sessionStartedAt: '2026-07-04T00:00:00Z' } }, bodyweightOnly: false };
    const shown = formatPrMeasure(measureEntry(loadedPlank, 'heaviest'), 'heaviest', loadedPlank, 'lb');
    expect(shown.caption).toBe('Heaviest load held');
  });

  it('renders est. 1RM three ways, matching how the record is actually ranked', () => {
    expect(formatPrMeasure(measureEntry(loaded, 'est1rm'), 'est1rm', loaded, 'lb')).toEqual({
      value: '208 lb',
      caption: '185lb×5',
    });
    expect(formatPrMeasure(measureEntry(pullUp, 'est1rm'), 'est1rm', pullUp, 'lb')).toEqual({
      value: '12 reps',
      caption: 'Bodyweight',
    });
    expect(formatPrMeasure(measureEntry(plank, 'est1rm'), 'est1rm', plank, 'lb')).toEqual({
      value: '2:10',
      caption: 'Longest hold',
    });
  });

  it('converts to the household unit, so the number agrees with the ranking', () => {
    const shown = formatPrMeasure(measureEntry(loaded, 'heaviest'), 'heaviest', loaded, 'kg');
    expect(shown.value).toBe('102 kg');
  });
});

describe('measureUnavailableCaption', () => {
  it('names why, so a dash reads as an answer rather than a gap', () => {
    expect(measureUnavailableCaption(pullUp)).toBe('Bodyweight');
    expect(measureUnavailableCaption(plank)).toBe('Timed hold');
    expect(measureUnavailableCaption({})).toBe('Not recorded');
  });

  // The user-visible point of the breakdown: a volume record with no detail is unreadable, and
  // unreadable is what let ten junk sets of an empty bar look identical to a genuine heavy day.
  describe('naming the work behind a session-level record', () => {
    function withBreakdown(sets, setCount) {
      return {
        exerciseId: 1,
        exerciseName: 'Squat',
        best: { weight: 185, reps: 5, unit: 'lb', est1rm: 215.8, sessionStartedAt: '2026-07-01T12:00:00Z' },
        measures: {
          sessionVolume: {
            value: 4000,
            weightLb: null,
            reps: null,
            sessionStartedAt: '2026-07-01T12:00:00Z',
            sets,
            setCount,
          },
        },
        bodyweightOnly: false,
        durationTracked: false,
      };
    }

    it('lists the runs instead of saying "One session"', () => {
      const row = withBreakdown(
        [
          { weightLb: 135, reps: 10, durationSeconds: null, count: 1 },
          { weightLb: 155, reps: 8, durationSeconds: null, count: 3 },
        ],
        4,
      );
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'lb').caption).toBe(
        '135lb×10, 3×155lb×8',
      );
    });

    // ⚠️ EVERY run the server sent is rendered. There used to be a second, tighter client cap of
    // three runs, whose "+N more" tail was honest about the count but pointed at sets there was no
    // way to see: the row is one button that opens a destination chooser, so the tail was not
    // tappable and nothing else revealed them. Showing the work IS the feature -- it is what
    // distinguishes a real heavy day from ten junk sets of an empty bar.
    it('shows every run the server sent, with no client-side cap', () => {
      const row = withBreakdown(
        [
          { weightLb: 135, reps: 10, durationSeconds: null, count: 1 },
          { weightLb: 155, reps: 8, durationSeconds: null, count: 1 },
          { weightLb: 165, reps: 8, durationSeconds: null, count: 1 },
          { weightLb: 175, reps: 6, durationSeconds: null, count: 5 },
        ],
        8,
      );
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'lb').caption).toBe(
        '135lb×10, 155lb×8, 165lb×8, 5×175lb×6',
      );
    });

    // ⚠️ "+N more" SURVIVES, and now means the only thing it can: the SERVER truncated
    // (MAX_PR_BREAKDOWN_RUNS). setCount is the true total, so the tail never understates the work
    // -- and it still counts SETS, not runs, or a row whose one visible run collapses eight sets
    // would claim far less than it holds.
    it('still tails honestly when the server itself truncated, counting SETS not runs', () => {
      const row = withBreakdown(
        [
          { weightLb: 135, reps: 10, durationSeconds: null, count: 1 },
          { weightLb: 155, reps: 8, durationSeconds: null, count: 3 },
        ],
        9, // the server sent 2 runs covering 4 sets, but the session really held 9
      );
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'lb').caption).toBe(
        '135lb×10, 3×155lb×8 +5 more',
      );
    });

    it('converts to the household unit like every other number on the board', () => {
      const row = withBreakdown([{ weightLb: 220.5, reps: 5, durationSeconds: null, count: 2 }], 2);
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'kg').caption).toBe(
        '2×100kg×5',
      );
    });

    it('falls back to "One session" when the row predates the breakdown', () => {
      const row = withBreakdown(undefined, 0);
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'lb').caption).toBe(
        'One session',
      );
    });

    // Reps IS available for a bodyweight exercise, so this is a real row, not a hypothetical.
    // "0lb" would be the column-of-zeros mistake, and keeping the "x" separator would render two
    // sets of twelve as the nonsense "2xx12".
    it('drops the weight for a bodyweight run rather than printing "0lb"', () => {
      const row = withBreakdown([{ weightLb: 0, reps: 12, durationSeconds: null, count: 2 }], 2);
      expect(formatPrMeasure(measureEntry(row, 'sessionVolume'), 'sessionVolume', row, 'lb').caption).toBe('2×12');
    });

    it('renders a hold as a time, with its load only when there is one', () => {
      const unloaded = withBreakdown([{ weightLb: 0, reps: 0, durationSeconds: 60, count: 3 }], 3);
      expect(formatPrMeasure(measureEntry(unloaded, 'sessionVolume'), 'sessionVolume', unloaded, 'lb').caption).toBe(
        '3×1:00',
      );
      const loaded = withBreakdown([{ weightLb: 25, reps: 0, durationSeconds: 60, count: 1 }], 1);
      expect(formatPrMeasure(measureEntry(loaded, 'sessionVolume'), 'sessionVolume', loaded, 'lb').caption).toBe(
        '25lb 1:00',
      );
    });
  });
});

File: backend/src/test/java/com/worktrac/backend/PrMeasuresTest.java
package com.worktrac.backend;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.worktrac.backend.email.EmailService;
import com.worktrac.backend.support.AbstractIntegrationTest;
import com.worktrac.backend.support.RegistrationTestSupport;
import com.worktrac.backend.user.TestCodeCache;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.http.MediaType;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

// The four measures the PRs board offers beside est. 1RM.
//
// The point of most of these is that the measures DISAGREE with each other -- a heavy single tops
// the bar but loses on est. 1RM, and the session holding the best single set is not necessarily the
// session with the most total volume. A fixture where they all agree would pass just as happily
// against an implementation that returned the same set for every measure, which is exactly the
// regression worth guarding against.
@AutoConfigureMockMvc
class PrMeasuresTest extends AbstractIntegrationTest {

    @DynamicPropertySource
    static void datasource(DynamicPropertyRegistry registry) {
        registerDatasource(registry, PrMeasuresTest.class);
    }

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private TestCodeCache testCodeCache;

    @MockitoBean
    private EmailService emailService;

    private final ObjectMapper objectMapper = new ObjectMapper();

    private String token;
    private long personId;
    private long barbellId;

    @BeforeEach
    void setUp() throws Exception {
        String email = "measures-" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        JsonNode registration = RegistrationTestSupport.registerAndConfirm(
                mockMvc, objectMapper, testCodeCache, email, "Nate");
        token = registration.get("token").asText();
        personId = registration.get("person").get("id").asLong();
        barbellId = createExercise("Measured Bench " + UUID.randomUUID(), "strength");
    }

    @Test
    void everyMeasureIsReportedForALoadedLift() throws Exception {
        // Session A: 100x10 then 100x10  -> volume 2000, reps 20, best set 1000, est1rm 100*(1+10/30)
        long a = createPastSession(daysAgo(20));
        logSet(a, barbellId, 100, 10);
        logSet(a, barbellId, 100, 10);
        // Session B: one heavy single. Top weight lives HERE, but volume and reps do not.
        long b = createPastSession(daysAgo(13));
        logSet(b, barbellId, 200, 1);

        JsonNode measures = rowFor(barbellId).get("measures");

        // Top weight is the heavy single, from session B.
        assertEquals(200.0, measures.get("heaviest").get("value").asDouble(), 0.01);
        assertEquals(1, measures.get("heaviest").get("reps").asInt());
        // Best SESSION volume is session A's 2000, not session B's 200.
        assertEquals(2000.0, measures.get("sessionVolume").get("value").asDouble(), 0.01);
        // Best SINGLE SET volume is 100x10 = 1000, not the session's 2000.
        assertEquals(1000.0, measures.get("bestSetVolume").get("value").asDouble(), 0.01);
        assertEquals(100.0, measures.get("bestSetVolume").get("weightLb").asDouble(), 0.01);
        assertEquals(10, measures.get("bestSetVolume").get("reps").asInt());
        // Best SESSION rep total is A's 20.
        assertEquals(20.0, measures.get("totalReps").get("value").asDouble(), 0.01);
    }

    // The distinction the "session total, not one set" copy exists for. If sessionVolume were
    // computed per set it would report 1000 here, and if bestSetVolume were computed per session it
    // would report 2000 -- this fixture catches either mistake.
    @Test
    void sessionLevelMeasuresPickTheSessionRatherThanTheSet() throws Exception {
        // The session with the best SINGLE SET is deliberately NOT the session with the most volume.
        long heavySingleSet = createPastSession(daysAgo(20));
        logSet(heavySingleSet, barbellId, 300, 5); // one set, volume 1500

        long grindySession = createPastSession(daysAgo(13));
        logSet(grindySession, barbellId, 100, 10); // 1000
        logSet(grindySession, barbellId, 100, 10); // 1000
        logSet(grindySession, barbellId, 100, 10); // 1000 -> session volume 3000, reps 30

        JsonNode measures = rowFor(barbellId).get("measures");

        assertEquals(1500.0, measures.get("bestSetVolume").get("value").asDouble(), 0.01);
        assertEquals(3000.0, measures.get("sessionVolume").get("value").asDouble(), 0.01);
        assertEquals(30.0, measures.get("totalReps").get("value").asDouble(), 0.01);
        // A session-level record names no single set -- no one set is the answer.
        assertTrue(measures.get("sessionVolume").get("weightLb").isNull());
        assertTrue(measures.get("sessionVolume").get("reps").isNull());
        assertTrue(measures.get("totalReps").get("weightLb").isNull());

        // ...but it DOES name the whole session's work. "One session" on its own was unreadable:
        // there was no way to tell a genuine heavy day from ten junk sets of an empty bar, which
        // is precisely how a volume record can be gamed.
        JsonNode breakdown = measures.get("sessionVolume").get("sets");
        assertEquals(1, breakdown.size(), "three identical sets collapse into one run");
        assertEquals(100.0, breakdown.get(0).get("weightLb").asDouble(), 0.01);
        assertEquals(10, breakdown.get(0).get("reps").asInt());
        assertEquals(3, breakdown.get(0).get("count").asInt());
        // The TRUE number of sets, so a truncated list can be labelled honestly.
        assertEquals(3, measures.get("sessionVolume").get("setCount").asInt());
    }

    // A set-level measure already names its own set through weightLb/reps, so it carries no
    // breakdown -- two representations of one thing is what PrRowDto's own header warns about.
    @Test
    void setLevelMeasuresCarryNoBreakdown() throws Exception {
        long session = createPastSession(daysAgo(20));
        logSet(session, barbellId, 225, 5);

        JsonNode measures = rowFor(barbellId).get("measures");

        assertEquals(0, measures.get("heaviest").get("sets").size());
        assertEquals(0, measures.get("heaviest").get("setCount").asInt());
        assertEquals(0, measures.get("bestSetVolume").get("sets").size());
    }

    // Consecutive identical sets collapse; a change of weight or reps starts a new run. Order is
    // chronological, so a ramp reads the way it was actually performed.
    @Test
    void theBreakdownCollapsesRunsAndKeepsChronologicalOrder() throws Exception {
        long session = createPastSession(daysAgo(10));
        logSet(session, barbellId, 135, 10);
        logSet(session, barbellId, 155, 8);
        logSet(session, barbellId, 155, 8);
        logSet(session, barbellId, 175, 6);

        JsonNode breakdown = rowFor(barbellId).get("measures").get("sessionVolume").get("sets");

        assertEquals(3, breakdown.size());
        assertEquals(135.0, breakdown.get(0).get("weightLb").asDouble(), 0.01);
        assertEquals(1, breakdown.get(0).get("count").asInt());
        assertEquals(155.0, breakdown.get(1).get("weightLb").asDouble(), 0.01);
        assertEquals(2, breakdown.get(1).get("count").asInt());
        assertEquals(175.0, breakdown.get(2).get("weightLb").asDouble(), 0.01);
    }

    // ⚠️ The cap exists because this DTO rides on the PRs board -- one row per exercise -- which
    // offlineCacheWarm persists to IndexedDB. setCount must still report the truth, or a long
    // workout would be shown as a shorter one.
    @Test
    void theBreakdownIsCappedButSetCountStaysHonest() throws Exception {
        long session = createPastSession(daysAgo(5));
        // Ten distinct runs, comfortably past the six-run cap.
        for (int i = 0; i < 10; i++) {
            logSet(session, barbellId, 100 + i * 5, 5);
        }

        JsonNode measure = rowFor(barbellId).get("measures").get("sessionVolume");

        assertEquals(6, measure.get("sets").size(), "capped at MAX_PR_BREAKDOWN_RUNS");
        assertEquals(10, measure.get("setCount").asInt(), "but the true total is still reported");
    }

    // Weight-derived measures must be ABSENT, not zero, for an exercise that was never loaded --
    // a column of "0 lb" is worse than no column. Reps survive: they are its honest record.
    @Test
    void weightMeasuresAreNullForABodyweightOnlyExerciseButRepsSurvive() throws Exception {
        long pullUpId = createExercise("Measured Pull-Up " + UUID.randomUUID(), "strength");
        long session = createPastSession(daysAgo(20));
        logSet(session, pullUpId, 0, 12);
        logSet(session, pullUpId, 0, 8);

        JsonNode row = rowFor(pullUpId);
        JsonNode measures = row.get("measures");

        assertTrue(row.get("bodyweightOnly").asBoolean());
        assertTrue(measures.get("heaviest").isNull());
        assertTrue(measures.get("sessionVolume").isNull());
        assertTrue(measures.get("bestSetVolume").isNull());
        assertEquals(20.0, measures.get("totalReps").get("value").asDouble(), 0.01);
    }

    // A hold carries reps = 0, so both volume measures and the rep total collapse to zero the same
    // way a never-loaded exercise collapses the weight ones. Added load still gives a top weight --
    // that is the "heaviest load held" record.
    @Test
    void repAndVolumeMeasuresAreNullForAHoldButAddedLoadStillCounts() throws Exception {
        long plankId = createExercise("Measured Plank " + UUID.randomUUID(), "duration");
        long session = createPastSession(daysAgo(20));
        logHold(session, plankId, 25, 90);
        logHold(session, plankId, 10, 130);

        JsonNode row = rowFor(plankId);
        JsonNode measures = row.get("measures");

        assertTrue(row.get("durationTracked").asBoolean());
        assertTrue(measures.get("totalReps").isNull());
        assertTrue(measures.get("sessionVolume").isNull());
        assertTrue(measures.get("bestSetVolume").isNull());
        assertEquals(25.0, measures.get("heaviest").get("value").asDouble(), 0.01);
    }

    // An unloaded hold has nothing on any of the four. est. 1RM (row.best) still carries its
    // duration, which is the whole reason that one is not inside `measures`.
    @Test
    void anUnloadedHoldHasNoMeasuresAtAllButKeepsItsDuration() throws Exception {
        long plankId = createExercise("Measured Floor Plank " + UUID.randomUUID(), "duration");
        long session = createPastSession(daysAgo(20));
        logHold(session, plankId, 0, 120);

        JsonNode row = rowFor(plankId);

        assertTrue(row.get("measures").get("heaviest").isNull());
        assertTrue(row.get("measures").get("sessionVolume").isNull());
        assertTrue(row.get("measures").get("bestSetVolume").isNull());
        assertTrue(row.get("measures").get("totalReps").isNull());
        assertEquals(120, row.get("best").get("durationSeconds").asInt());
    }

    // Weights are normalized to POUNDS server-side, not left in the set's own unit. That is what
    // lets the client rank and render every measure through one path; prSort.js otherwise has to
    // call toLb per measure, which is exactly the per-call-site derivation that drifts. A set's
    // unit comes from the account default, so this switches the household to kg first.
    @Test
    void measuresAreNormalizedToPounds() throws Exception {
        setDefaultUnit("kg");
        long session = createPastSession(daysAgo(20));
        logSet(session, barbellId, 100, 5);

        JsonNode measures = rowFor(barbellId).get("measures");

        // 100 kg is 220.46 lb, and 100 kg x 5 is 1102.31 lb of volume -- not 100 and 500.
        assertEquals(220.5, measures.get("heaviest").get("value").asDouble(), 0.1);
        assertEquals(1102.3, measures.get("bestSetVolume").get("value").asDouble(), 0.5);
        // The set itself is untouched: only the derived measure is converted.
        assertEquals(100.0, rowFor(barbellId).get("best").get("weight").asDouble(), 0.01);
    }

    // Relative to now, not a date literal. A brand-new account is on Free, whose 90-day window is a
    // read filter on getPrList -- a fixed 2026-05-01 fixture would pass today and silently start
    // returning an empty board once the real clock moved past the window.
    private String daysAgo(int days) {
        return Instant.now().minus(Duration.ofDays(days)).truncatedTo(ChronoUnit.SECONDS).toString();
    }

    private long createExercise(String name, String trackingType) throws Exception {
        Map<String, Object> body = new HashMap<>();
        body.put("name", name);
        body.put("trackingType", trackingType);
        String response = mockMvc.perform(post("/api/exercises")
                        .header("Authorization", "Bearer " + token)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(body)))
                .andExpect(status().isOk())
                .andReturn().getResponse().getContentAsString();
        return objectMapper.readTree(response).get("id").asLong();
    }

    private long createPastSession(String startedAt) throws Exception {
        String response = mockMvc.perform(post("/api/people/" + personId + "/sessions")
                        .header("Authorization", "Bearer " + token)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(Map.of("startedAt", startedAt))))
                .andExpect(status().isOk())
                .andReturn().getResponse().getContentAsString();
        return objectMapper.readTree(response).get("id").asLong();
    }

    private void setDefaultUnit(String unit) throws Exception {
        mockMvc.perform(org.springframework.test.web.servlet.request.MockMvcRequestBuilders
                        .put("/api/account/default-unit")
                        .header("Authorization", "Bearer " + token)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(Map.of("defaultUnit", unit))))
                .andExpect(status().isOk());
    }

    private void logSet(long sessionId, long exerciseId, double weight, int reps) throws Exception {
        Map<String, Object> body = new HashMap<>();
        body.put("exerciseId", exerciseId);
        body.put("weight", weight);
        body.put("reps", reps);
        mockMvc.perform(post("/api/sessions/" + sessionId + "/sets")
                        .header("Authorization", "Bearer " + token)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(body)))
                .andExpect(status().isOk());
    }

    private void logHold(long sessionId, long exerciseId, double weight, int durationSeconds) throws Exception {
        Map<String, Object> body = new HashMap<>();
        body.put("exerciseId", exerciseId);
        body.put("weight", weight);
        // reps is 0 on a hold, never absent -- a hold genuinely has zero repetitions, and the
        // column is NOT NULL so every rep-derived aggregate stays correct with no null handling.
        body.put("reps", 0);
        body.put("durationSeconds", durationSeconds);
        mockMvc.perform(post("/api/sessions/" + sessionId + "/sets")
                        .header("Authorization", "Bearer " + token)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(body)))
                .andExpect(status().isOk());
    }

    private JsonNode rowFor(long exerciseId) throws Exception {
        String response = mockMvc.perform(get("/api/people/" + personId + "/prs")
                        .header("Authorization", "Bearer " + token))
                .andExpect(status().isOk())
                .andReturn().getResponse().getContentAsString();
        for (JsonNode row : objectMapper.readTree(response)) {
            if (row.get("exerciseId").asLong() == exerciseId) return row;
        }
        throw new AssertionError("No PR row for exercise " + exerciseId + " in " + response);
    }
}
